### PR TITLE
Issue#71 Introduce Query Service

### PR DESCRIPTION
**Pull request by OlTym**
dev
## ZenHub

* [Main ZenHub ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009)

## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: home-api-tests/pom.xml [hunk 1]
```diff
@@ -34,6 +34,10 @@
         <liquibase.internal.port>8000</liquibase.internal.port>
         <home.application.internal.port>8080</home.application.internal.port>
         <docker.artifacts.path>${project.build.directory}/docker/artifacts</docker.artifacts.path>
+        <commons-text-version>1.9</commons-text-version>
+        <rsql-common-version>5.0.6</rsql-common-version>
+
```
Comment (kh0ma) [inline home-api-tests/pom.xml]: Remove extra lines

File: home-api-tests/pom.xml [hunk 2]
```diff
@@ -258,8 +262,10 @@
                                 </home.application.external.port>
                                 <postgres.external.port>${postgres.external.port}</postgres.external.port>
                                 <liquibase.external.port>${liquibase.external.port}</liquibase.external.port>
-                                <home.application.admin.username>${home.application.admin.username}</home.application.admin.username>
-                                <home.application.admin.password>${home.application.admin.password}</home.application.admin.password>
+                                <home.application.admin.username>${home.application.admin.username}
+                                </home.application.admin.username>
```
Comment (kh0ma) [inline home-api-tests/pom.xml]: Why did you change it?

File: home-api-tests/pom.xml [hunk 3]
```diff
@@ -258,8 +262,10 @@
...
+                                <home.application.admin.password>${home.application.admin.password}
```
Comment (kh0ma) [inline home-api-tests/pom.xml]: Why did you change it?

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/NewsApiIT.java [hunk 4]
```diff
@@ -1,22 +1,23 @@
-package com.softserveinc.ita.homeproject.api.tests;
-
-import java.util.List;
-
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Test;
+package com.softserveinc.ita.homeproject.api.tests.NewsApiTests;
```
Comment (kh0ma) [inline home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/NewsApiIT.java]: Package name doesn't match common naming style

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/NewsApiIT.java [hunk 5]
```diff
@@ -59,15 +60,6 @@ void updateNewsTest() throws ApiException {
         assertNews(savedNews, updateNews, updatedNews);
     }
 
-    @Test
```
Comment (kh0ma) [inline home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/NewsApiIT.java]: Why did you remove it? This test is still relevant

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/QueryNewsIT.java [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+    private final CreateNews expectedNews = new CreateNews()
+            .description("description")
+            .source("source")
+            .text("text")
+            .photoUrl("photoUrl");
+
+    @BeforeEach
+    public void setUp() {
+        newsApi = new NewsApi(ApiClientUtil.getClient());
```
Comment (kh0ma) [inline home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/QueryNewsIT.java]: Move it to field declaration

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/QueryNewsIT.java [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        expectedNews.setTitle(RandomStringUtils.randomAlphabetic(7));
```
Comment (kh0ma) [inline home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/QueryNewsIT.java]: Move it to field declaration

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/QueryNewsIT.java [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    }
+
+    @Test
+    void getAllNewsByTitleTest() throws ApiException {
+        newsApi.addNews(expectedNews);
+
+        List<ReadNews> actualListNews = new NewsQuery
+                .Builder(pageNumber, pageSize, newsApi)
```
Comment (kh0ma) [inline home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/QueryNewsIT.java]: I think pageSize and pageNumber should be as other query parameters. There is no need make it required.

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/QueryNewsIT.java [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+        actualListNews.forEach(readNews -> assertThat(readNews.getTitle()).isEqualTo(expectedNews.getTitle()));
+    }
+
+    @Test
+    void getAllNewsBySourceTest() throws ApiException {
+        newsApi.addNews(expectedNews);
+
+        List<ReadNews> actualListNews = new NewsQuery
+                .Builder(pageNumber, pageSize, newsApi)
```
Comment (kh0ma) [inline home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/QueryNewsIT.java]: Please use method for getting a builder. I see it as following:
```java
new NewsQuery(newsApi).builder()
    .pageSize(10)
    .source("Dummy source)
    .build()
    .perform();
```

File: home-api-tests/pom.xml [hunk 10]
```diff
@@ -310,5 +316,16 @@
             <artifactId>commons-lang3</artifactId>
             <version>${commons-lang3-version}</version>
         </dependency>
+        <dependency>
+            <groupId>io.github.perplexhub</groupId>
+            <artifactId>rsql-common</artifactId>
```
Comment (kh0ma) [inline home-api-tests/pom.xml]: Can we get rid of this dependency in tests?

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/UserApiImpl.java [hunk 11]
```diff
@@ -28,10 +36,11 @@
  */
 @Provider
 @NoArgsConstructor
-public class UserApiImpl implements UsersApi {
+public class UserApiImpl extends CommonApi<UserDto> implements UsersApi {
 
     private UserService userService;
     private HomeMapper mapper;
+    private EntitySpecificationService entitySpecificationService;
```
Comment (kh0ma) [inline home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/UserApiImpl.java:43]: Can we move mapper and entitySpecificationService into CommonApi?

File: home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/repository/NewsRepository.java [hunk 12]
```diff
@@ -1,15 +1,11 @@
 package com.softserveinc.ita.homeproject.homedata.repository;
 
 import com.softserveinc.ita.homeproject.homedata.entity.News;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
 import org.springframework.data.repository.PagingAndSortingRepository;
 import org.springframework.stereotype.Repository;
 
-/**
- * NewsRepository is the interface that is needed
```
Comment (kh0ma) [inline home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/repository/NewsRepository.java:8]: Why did you remove this doc?

File: home-data/src/main/resources/application.properties [hunk 13]
```diff
@@ -3,3 +3,4 @@ spring.datasource.username=${DATASOURCE_USER:postgres}
 spring.datasource.password=${DATASOURCE_PASSWORD:password}
 spring.datasource.driver-class-name=org.postgresql.Driver
 spring.jpa.hibernate.ddl-auto=validate
+
```
Comment (kh0ma) [inline home-data/src/main/resources/application.properties]: Remove extra line

File: home-data/pom.xml [hunk 14]
```diff
@@ -16,9 +16,18 @@
...
 
     <dependencies>
+
+        <dependency>
+            <groupId>io.github.perplexhub</groupId>
+            <artifactId>rsql-jpa-spring-boot-starter</artifactId>
+            <version>${rsql-jpa-spring-boot-starter.version}</version>
+        </dependency>
+
+
```
Comment (kh0ma) [inline home-data/pom.xml]: Remove extralines

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/query/EntitySpecificationService.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    public static final String DEFAULT_SORT = "id,desc";
+    public static final String RSQL_EQUAL = "==";
+    public static final String RSQL_AND = ";";
+
+    public Specification<T> getSpecification(Map<QueryParamEnum, String> filter, String search, String sort) {
+        Specification<T> filterSpecification = RSQLJPASupport.toSpecification(toRQSQLString(filter));
+        Specification<T> searchSpecification = RSQLJPASupport.toSpecification(search);
+        Specification<T> sortSpecification = RSQLJPASupport.toSort(sort == null ? DEFAULT_SORT : sort);
+
+        return sortSpecification.and(filterSpecification).and(searchSpecification);
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/query/EntitySpecificationService.java]: Fix spotbug issue https://sonarcloud.io/project/issues?id=ita-social-projects-home&open=AXeROB1rnNgMB2oNji1b&pullRequest=167&resolved=false&types=BUG

**Review by kh0ma — CHANGES_REQUESTED** [review]

File: home-api-tests/pom.xml (hunk 1 above)
Comment (OlTym) [reply home-api-tests/pom.xml]: done

File: home-api-tests/pom.xml (hunk 2 above)
Comment (OlTym) [reply home-api-tests/pom.xml]: insert \r\n

File: home-api-tests/pom.xml (hunk 3 above)
Comment (OlTym) [reply home-api-tests/pom.xml]: insert \r\n

File: home-api-tests/pom.xml (hunk 10 above)
Comment (OlTym) [reply home-api-tests/pom.xml]: done

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/NewsApiIT.java (hunk 4 above)
Comment (OlTym) [reply home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/NewsApiIT.java]: done

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/NewsApiIT.java (hunk 5 above)
Comment (OlTym) [reply home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/NewsApiIT.java]: done

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/QueryNewsIT.java (hunk 6 above)
Comment (OlTym) [reply home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/NewsApiTests/QueryNewsIT.java]: done

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/UserApiImpl.java (hunk 11 above)
Comment (OlTym) [reply home-application/src/main/java/com/softserveinc/ita/homeproject/application/api/UserApiImpl.java:43]: If we move  mapper and entitySpecificationService into CommonApi We'll get NullPointerException

File: home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/repository/NewsRepository.java (hunk 12 above)
Comment (OlTym) [reply home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/repository/NewsRepository.java:8]: this doc is inserted

File: home-data/src/main/resources/application.properties (hunk 13 above)
Comment (OlTym) [reply home-data/src/main/resources/application.properties]: done

File: home-api-tests/pom.xml (hunk 2 above)
Comment (kh0ma) [reply home-api-tests/pom.xml]: Please remove your insertion

File: home-api-tests/pom.xml (hunk 3 above)
Comment (kh0ma) [reply home-api-tests/pom.xml]: Please remove your insertion

File: home-api-tests/pom.xml [hunk 16]
```diff
@@ -310,5 +315,10 @@
             <artifactId>commons-lang3</artifactId>
             <version>${commons-lang3-version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-text</artifactId>
+            <version>${commons-text-version}</version>
```
Comment (kh0ma) [inline home-api-tests/pom.xml:237]: IIRC, `org.apache.commons` distributes all the libraries in the same version. Please make it common for `commons-text` and `commons-lang3`

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+import cz.jirutka.rsql.parser.RSQLParserException;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.dao.InvalidDataAccessApiUsageException;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+public final class ExceptionMapperUtils {
+
+    public static String RSQLExceptionMessageParser(RSQLParserException exception) {
```
Comment (kh0ma) [inline home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java]: naming

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+                } else {
+                    message = result.get().trim();
+                }
+            }
+        }
+        return message;
+    }
+
+    private static StringBuilder getProperty(String result) {
+             StringBuilder  sb = new StringBuilder(result);
```
Comment (kh0ma) [inline home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java]: I see here some problems with indentation.

File: home-data/pom.xml [hunk 19]
```diff
@@ -16,9 +16,17 @@
 
     <properties>
         <postgresql.verson>42.2.16</postgresql.verson>
+        <rsql-jpa-spring-boot-starter.version>5.0.6</rsql-jpa-spring-boot-starter.version>
+         <commons-lang3.version>3.11</commons-lang3.version>
```
Comment (kh0ma) [inline home-data/pom.xml]: indentation

**Review by kh0ma — CHANGES_REQUESTED** [review]

**Review by alexorlenko — APPROVED** [review]

File: home-api-tests/pom.xml (hunk 16 above)
Comment (OlTym) [reply home-api-tests/pom.xml:237]: The last version of the Apache Commons Text library is 1.9. The last version of the  Apache Commons Lang library is 3.11 The versions of these libraries are not same You can compare  versions of this library using the following links
https://mvnrepository.com/artifact/org.apache.commons/commons-text
https://mvnrepository.com/artifact/org.apache.commons/commons-lang3

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java (hunk 18 above)
Comment (OlTym) [reply home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java]: done

File: home-data/pom.xml (hunk 19 above)
Comment (OlTym) [reply home-data/pom.xml]: done

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java (hunk 17 above)
Comment (OlTym) [reply home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java]: done

File: home-api-tests/pom.xml (hunk 16 above)
Comment (kh0ma) [reply home-api-tests/pom.xml:237]: Please find `org.apache.commons.lang3.StringEscapeUtils` It looks like we don't need `commons-text` at all.

**Review by kh0ma — CHANGES_REQUESTED** [review]

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+import cz.jirutka.rsql.parser.RSQLParserException;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.dao.InvalidDataAccessApiUsageException;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+public final class ExceptionMapperUtils {
+
+    public static String BadRequestExceptionMessageParser(RSQLParserException exception) {
```
Comment (kh0ma) [inline home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java]: Please pay attention to the method naming!

**Review by kh0ma — CHANGES_REQUESTED** [review]

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java (hunk 17 above)
Comment (kh0ma) [reply home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java]: Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: home-api-tests/pom.xml (hunk 16 above)
Comment (OlTym) [reply home-api-tests/pom.xml:237]: The method StringEscapeUtils is depricated in the org.apache.commons.lang3 becouse I used method StringEscapeUtils from commons-text library

Comment (kh0ma) [reply home-api-tests/pom.xml:237]: Using older versions for avoiding deprecation notices is not the way how should deal with it. If something is deprecated we should find an actual alternative. Please identify why this class/method is deprecated in the new version of the library.

Comment (kh0ma) [reply home-api-tests/pom.xml:237]: Ok. I got it. The `commons-text` is another library that apache support. They move all the text stuff there. So It's ok. Relosved.

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java (hunk 20 above)
Comment (OlTym) [reply home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java]: done

**Review by kh0ma — APPROVED** [review]

**Review by PoLytvy — APPROVED** [review]

**Review by Romanov-Niko — APPROVED** [review]

**Review by VadymSokorenko — APPROVED** [review]
